code/modules/reagents/chemistry/holder.dm

```diff
@@ -677,138 +727,320 @@
-			var/list/cached_results = selected_reaction.results
-			var/list/multiplier = INFINITY
-			for(var/B in cached_required_reagents)
-				multiplier = min(multiplier, round(get_reagent_amount(B) / cached_required_reagents[B]))
+					reaction_list += E
 
-			for(var/B in cached_required_reagents)
-				remove_reagent(B, (multiplier * cached_required_reagents[B]), safety = 1)
+	if(reaction_list.len)
+		is_reacting = TRUE //We've entered the reaction phase - this is set here so any reagent handling called in on_reaction() doesn't cause infinite loops
+		START_PROCESSING(SSprocessing, src) //see process() to see how reactions are handled
+	else
+		is_reacting = FALSE
 
-			for(var/P in selected_reaction.results)
-				multiplier = max(multiplier, 1) //this shouldn't happen ...
-				SSblackbox.record_feedback("tally", "chemical_reaction", cached_results[P]*multiplier, P)
-				add_reagent(P, cached_results[P]*multiplier, null, chem_temp)
+	if(.)
+		SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
 
-			var/list/seen = viewers(4, get_turf(my_atom))
-			var/iconhtml = icon2html(cached_my_atom, seen)
-			if(cached_my_atom)
-				if(!ismob(cached_my_atom)) // No bubbling mobs
-					if(selected_reaction.mix_sound)
-						playsound(get_turf(cached_my_atom), selected_reaction.mix_sound, 80, TRUE)
-
-					for(var/mob/M in seen)
-						to_chat(M, "<span class='notice'>[iconhtml] [selected_reaction.mix_message]</span>")
-
-				if(istype(cached_my_atom, /obj/item/slime_extract))
-					var/obj/item/slime_extract/ME2 = my_atom
-					ME2.Uses--
-					if(ME2.Uses <= 0) // give the notification that the slime core is dead
-						for(var/mob/M in seen)
-							to_chat(M, "<span class='notice'>[iconhtml] \The [my_atom]'s power is consumed in the reaction.</span>")
-							ME2.name = "used slime extract"
-							ME2.desc = "This extract has been used up."
-
-			selected_reaction.on_reaction(src, multiplier)
-			reaction_occurred = TRUE
-			.++
+/*
+* Main Reaction loop handler, Do not call this directly
+*
+* Checks to see if there's a reaction, then processes over the reaction list, removing them if flagged
+* If any are ended, it displays the reaction message and removes it from the reaction list
+* If the list is empty at the end it finishes reacting.
+* Arguments:
+* * delta_time - the time between each time step
+*/
+/datum/reagents/process(delta_time)
+	if(!is_reacting)
+		finish_reacting()
+	//sum of output messages.
+	var/mix_message
+	//Process over our reaction list
+	//See equilibrium.dm for mechanics
+	for(var/datum/equilibrium/E in reaction_list)
+		SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[E.reaction.type] total reaction steps")
+		//if it's been flagged to delete
+		if(E.to_delete)
+			mix_message += end_reaction(E)
+			continue
+		//otherwise continue reacting
+		E.react_timestep(delta_time)
+
+	if(mix_message)
+		my_atom.visible_message("<span class='notice'>[icon2html(my_atom, viewers(DEFAULT_MESSAGE_RANGE, src))] [mix_message]</span>")
```

Comment:
```suggestion
	var/mix_message = list()
	//Process over our reaction list
	//See equilibrium.dm for mechanics
	for(var/e in reaction_list)
		var/datum/equilibrium/E = e
		SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[E.reaction.type] total reaction steps")
		//if it's been flagged to delete
		if(E.to_delete)
			mix_message += end_reaction(E)
			continue
		//otherwise continue reacting
		E.react_timestep(delta_time)

	if(length(mix_message))
		my_atom.visible_message("<span class='notice'>[icon2html(my_atom, viewers(DEFAULT_MESSAGE_RANGE, src))] [mix_message.Join()]</span>")
```